internal/services/dns/dns_ds_record_resource.go

```diff
@@ -0,0 +1,328 @@
+			if err != nil {
+				return err
+			}
+
+			existing, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			if existing.Model == nil {
+				return fmt.Errorf("retrieving %s: `model` was nil", id)
+			}
+
+			if existing.Model.Properties == nil {
+				return fmt.Errorf("retrieving %s: `properties` was nil", id)
+			}
+
+			payload := *existing.Model
+
+			if metadata.ResourceData.HasChange("record") {
+				payload.Properties.DSRecords = expandAzureRmDnsDSRecords(model.Record)
+			}
+
+			if metadata.ResourceData.HasChange("ttl") {
+				payload.Properties.TTL = pointer.To(model.Ttl)
+			}
+
+			if metadata.ResourceData.HasChange("tags") {
+				payload.Properties.Metadata = pointer.To(model.Tags)
+			}
+
+			if _, err := client.CreateOrUpdate(ctx, *id, payload, recordsets.DefaultCreateOrUpdateOperationOptions()); err != nil {
+				return fmt.Errorf("updating %s: %+v", id, err)
+			}
+
+			return nil
+		},
+	}
+}
+
+func (DnsDSRecordResource) Delete() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Dns.RecordSets
+
+			id, err := recordsets.ParseRecordTypeID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			if _, err := client.Delete(ctx, *id, recordsets.DefaultDeleteOperationOptions()); err != nil {
+				return fmt.Errorf("deleting %s: %+v", *id, err)
+			}
+
+			return nil
+		},
+	}
+}
+
+func flattenAzureRmDnsDSRecords(records *[]recordsets.DsRecord) []DnsDSRecordResourceRecord {
+	results := make([]DnsDSRecordResourceRecord, 0)
+
+	if records != nil {
+		for _, record := range *records {
+			result := DnsDSRecordResourceRecord{
+				Algorithm: pointer.From(record.Algorithm),
+				KeyTag:    pointer.From(record.KeyTag),
+			}
+
+			if record.Digest != nil {
+				result.DigestType = pointer.From(record.Digest.AlgorithmType)
+				result.DigestValue = pointer.From(record.Digest.Value)
+			}
+			results = append(results, result)
+		}
+	}
+
+	return results
+}
+
+func expandAzureRmDnsDSRecords(d []DnsDSRecordResourceRecord) *[]recordsets.DsRecord {
+	records := make([]recordsets.DsRecord, 0)
+
+	for _, v := range d {
+		records = append(records, recordsets.DsRecord{
+			Algorithm: pointer.To(v.Algorithm),
+			KeyTag:    pointer.To(v.KeyTag),
+			Digest: &recordsets.Digest{
+				AlgorithmType: pointer.To(v.DigestType),
+				Value:         pointer.To(v.DigestValue),
+			},
+		})
+	}
+
+	return &records
+}
+
+func resourceDnsDSRecordHash(v interface{}) int {
+	var buf bytes.Buffer
+
+	if m, ok := v.(map[string]interface{}); ok {
+		buf.WriteString(fmt.Sprintf("%d-", m["algorithm"].(int)))
+		buf.WriteString(fmt.Sprintf("%d-", m["key_tag"].(int)))
+		buf.WriteString(fmt.Sprintf("%d-", m["digest_type"].(int)))
+		buf.WriteString(fmt.Sprintf("%s-", m["digest_value"].(string)))
+	}
+
+	return pluginsdk.HashString(buf.String())
+}
```

Comment:
```suggestion
```

Custom hash functions for sets are only necessary if a property in the set should not be included in the hash, e.g. computed properties whose values can be changed by the API should not be included in the hash so that Terraform doesn't incorrectly mark a set for update. Since we're building the hash using all the properties in the set we can probably remove this, as well as `Set: resourceDnsDSRecordHash` on line 99